### PR TITLE
Add SSN losses.

### PR DESCRIPTION
**Pull request by tony2016uestc**
This PR adds SSN losses which contain `OHEMHingeLoss`, `activity_loss`, `completeness_loss`, `classwise_regression_loss`.

### CODE REVIEW

File: mmaction/models/losses/OHEMHingeLoss.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    @staticmethod
+    def forward(ctx, pred, labels, is_positive, ohem_ratio, group_size):
+        num_samples = pred.size()[0]
+        if num_samples != len(labels):
+            raise ValueError(f'Number of samples should be equal to that'
+                             f'of labels, but got {num_samples} samples and'
+                             f'{len(labels)} labels.')
+
+        losses = torch.zeros(num_samples)
+        slopes = torch.zeros(num_samples)
```
Comment (dreamerlin) [inline mmaction/models/losses/OHEMHingeLoss.py:21]: `torch.zeros(num_samples, device=pred.device)`

File: mmaction/models/losses/OHEMHingeLoss.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        for i in range(num_samples):
+            losses[i] = max(0, 1 - is_positive * pred[i, labels[i] - 1])
+            slopes[i] = -is_positive if losses[i] != 0 else 0
+
+        losses = losses.view(-1, group_size).contiguous()
+        sorted_losses, indices = torch.sort(losses, dim=1, descending=True)
+        keep_length = int(group_size * ohem_ratio)
+        loss = torch.zeros(1)
```
Comment (dreamerlin) [inline mmaction/models/losses/OHEMHingeLoss.py:29]: `torch.zeros(1, device=pred.device)`

File: mmaction/models/losses/OHEMHingeLoss.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    """This class is the core implementation for the completeness loss in
+    paper.
+
+    It compute class-wise hinge loss and performs online hard negative mining
+    (OHEM).
+    """
+
+    @staticmethod
+    def forward(ctx, pred, labels, is_positive, ohem_ratio, group_size):
+        num_samples = pred.size()[0]
```
Comment (dreamerlin) [inline mmaction/models/losses/OHEMHingeLoss.py:14]: pred.size(0)

File: mmaction/models/losses/OHEMHingeLoss.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        ctx.group_size = group_size
+        ctx.num_groups = losses.size(0)
+        return loss
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        labels = ctx.labels
+        slopes = ctx.slopes
+
+        grad_in = torch.zeros(ctx.shape)
```
Comment (dreamerlin) [inline mmaction/models/losses/OHEMHingeLoss.py:45]: `torch.zeros(ctx.shape, device=ctx.losses.device)`

File: mmaction/models/losses/OHEMHingeLoss.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        for group in range(ctx.num_groups):
+            for idx in ctx.loss_index[group]:
+                loc = idx + group * ctx.group_size
+                grad_in[loc, labels[loc] - 1] = slopes[loc] * \
+                    grad_output.data[0]
+        return torch.autograd.Variable(grad_in.cuda()), None, None, None, None
```
Comment (dreamerlin) [inline mmaction/models/losses/OHEMHingeLoss.py:51]: `torch.autograd.Variable(grad_in)`

File: mmaction/models/losses/OHEMHingeLoss.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
+import torch
```
Comment (dreamerlin) [inline mmaction/models/losses/OHEMHingeLoss.py:1]: filename should be in lower case. `ohem_hinge_loss.py`

File: mmaction/models/losses/OHEMHingeLoss.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+class OHEMHingeLoss(torch.autograd.Function):
+    """This class is the core implementation for the completeness loss in
+    paper.
+
+    It compute class-wise hinge loss and performs online hard negative mining
+    (OHEM).
+    """
+
+    @staticmethod
+    def forward(ctx, pred, labels, is_positive, ohem_ratio, group_size):
```
Comment (dreamerlin) [inline mmaction/models/losses/OHEMHingeLoss.py:13]: missing args description.

File: mmaction/models/losses/ssn_loss.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+import torch.nn.functional as F
+
+from ..registry import LOSSES
+from .OHEMHingeLoss import OHEMHingeLoss
+
+
+@LOSSES.register_module()
+class SSNLoss(nn.Module):
+
+    def activity_loss(self, activity_score, labels, activity_indexer):
```
Comment (dreamerlin) [inline mmaction/models/losses/ssn_loss.py:12]: Miss args description.

File: mmaction/models/losses/ssn_loss.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        gt = labels[activity_indexer]
+        return F.cross_entropy(pred, gt)
+
+    def completeness_loss(self,
+                          completeness_score,
+                          labels,
+                          completeness_indexer,
+                          positive_per_video,
+                          incomplete_per_video,
+                          ohem_ratio=0.17):
```
Comment (dreamerlin) [inline mmaction/models/losses/ssn_loss.py:27]: Miss args description.

File: mmaction/models/losses/ssn_loss.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+            incomplete_pred, gt[:, positive_per_video:].contiguous().view(-1),
+            -1, ohem_ratio, incomplete_per_video)
+        num_positives = positive_pred.size(0)
+        num_incompletes = int(incomplete_pred.size()[0] * ohem_ratio)
+
+        return ((positive_loss + incomplete_loss) /
+                float(num_positives + num_incompletes))
+
+    def classwise_regression_loss(self, bbox_pred, labels, bbox_targets,
+                                  regression_indexer):
```
Comment (dreamerlin) [inline mmaction/models/losses/ssn_loss.py:58]: Miss args description.

File: mmaction/models/losses/ssn_loss.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+                          incomplete_per_video,
+                          ohem_ratio=0.17):
+        """Completeness Loss.
+
+        It will calculate completeness loss given completeness_score and label.
+        """
+        pred = completeness_score[completeness_indexer, :]
+        gt = labels[completeness_indexer]
+
+        pred_dim = pred.size()[1]
```
Comment (dreamerlin) [inline mmaction/models/losses/ssn_loss.py:35]: `pred_dim = pred.size(1)`

File: mmaction/models/losses/ssn_loss.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        # yapf:enable
+
+        positive_loss = OHEMHingeLoss.apply(
+            positive_pred, gt[:, :positive_per_video].contiguous().view(-1), 1,
+            1.0, positive_per_video)
+        incomplete_loss = OHEMHingeLoss.apply(
+            incomplete_pred, gt[:, positive_per_video:].contiguous().view(-1),
+            -1, ohem_ratio, incomplete_per_video)
+        num_positives = positive_pred.size(0)
+        num_incompletes = int(incomplete_pred.size()[0] * ohem_ratio)
```
Comment (dreamerlin) [inline mmaction/models/losses/ssn_loss.py:52]: `incomplete_pred.size(0)`

File: mmaction/models/losses/ssn_loss.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        classwise_reg_pred = torch.cat(
+            (torch.diag(classwise_pred[:, :, 0]).view(
+                -1, 1), torch.diag(classwise_pred[:, :, 1]).view(-1, 1)),
+            dim=1)
+        loss = F.smooth_l1_loss(
+            classwise_reg_pred.view(-1), reg_target.view(-1)) * 2
+        return loss
+
+    def forward(self, activity_score, completeness_score, bbox_pred,
+                proposal_type, labels, bbox_targets, train_cfg):
```
Comment (dreamerlin) [inline mmaction/models/losses/ssn_loss.py:79]: Miss args description.

File: mmaction/models/losses/ssn_loss.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+            self.sampler.num_per_video *
+            (self.sampler.background_ratio / total_ratio))
+        incomplete_per_video = (
+            self.sampler.num_per_video - positive_per_video -
+            background_per_video)
+
+        losses['loss_activity'] = self.activity_loss(activity_score, labels,
+                                                     activity_indexer),
+        losses['loss_activity'] = torch.tensor(
+            device=activity_score.device, requires_grad=True)
```
Comment (dreamerlin) [inline mmaction/models/losses/ssn_loss.py:107]: Miss `x` and this operation may be redundant once we early set the `device` and `requires_grad` when calculating the loss.

Comment (dreamerlin) [reply mmaction/models/losses/ssn_loss.py:107]: Same to `loss_completeness` and `loss_reg`.

File: tests/test_loss.py [hunk 15]
```diff
@@ -105,3 +105,37 @@ def test_bmn_loss():
     assert_array_almost_equal(output_loss[1].numpy(), output_tem_loss)
     assert_array_almost_equal(output_loss[2].numpy(), output_pem_reg_loss)
     assert_array_almost_equal(output_loss[3].numpy(), output_pem_cls_loss)
+
+
+def test_ssn_loss():
```
Comment (dreamerlin) [inline tests/test_loss.py:110]: Besides the sub function of the `SSNLoss`, you must also test the whole `SSNLoss` foward since it is the common case.

File: mmaction/models/losses/ohem_hinge_loss.py [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    @staticmethod
+    def forward(ctx, pred, labels, is_positive, ohem_ratio, group_size):
+        """Calculate OHEM hinge loss.
+
+        Args:
+            pred (torch.Tensor): Predicted completeness score.
+            labels (torch.Tensor): Groundtruth class label.
+            is_positive (int): Set to 1 when proposals are positive and
+                set to -1 when proposals are incomplete.
+            ohem_ratio (float): Groundtruth temporal_iou score for start.
```
Comment (dreamerlin) [inline mmaction/models/losses/ohem_hinge_loss.py:21]: The description seems not right.

File: mmaction/models/losses/ohem_hinge_loss.py [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+            group_size (int): Number of proposals sampled
+                per video.
```
Comment (dreamerlin) [inline mmaction/models/losses/ohem_hinge_loss.py:23]: Not need to write in a new line since it is not more than 80.

File: mmaction/models/losses/ohem_hinge_loss.py [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+    def backward(ctx, grad_output):
+        labels = ctx.labels
+        slopes = ctx.slopes
+
+        grad_in = torch.zeros(ctx.shape, device=ctx.losses.device)
+        for group in range(ctx.num_groups):
+            for idx in ctx.loss_index[group]:
+                loc = idx + group * ctx.group_size
+                grad_in[loc, labels[loc] - 1] = slopes[loc] * \
+                    grad_output.data[0]
```
Comment (dreamerlin) [inline mmaction/models/losses/ohem_hinge_loss.py:64]: bracket.

File: mmaction/models/losses/ssn_loss.py [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+            self.sampler.num_per_video *
+            (self.sampler.background_ratio / total_ratio))
+        incomplete_per_video = (
+            self.sampler.num_per_video - positive_per_video -
+            background_per_video)
+
+        losses['loss_activity'] = self.activity_loss(activity_score, labels,
+                                                     activity_indexer)
+        losses['loss_activity'] = torch.tensor(
+            losses['loss_activity'], device=activity_score.device)
```
Comment (dreamerlin) [inline mmaction/models/losses/ssn_loss.py:161]: This line is redundant since the return of `self.activity_loss()` already in activity_score device.

File: mmaction/models/losses/ssn_loss.py [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+        losses['loss_completeness'] = self.completeness_loss(
+            completeness_score,
+            labels,
+            completeness_indexer,
+            positive_per_video,
+            incomplete_per_video,
+            ohem_ratio=positive_per_video / incomplete_per_video)
+        losses['loss_completeness'] *= self.loss_weight.comp_loss_weight
+        losses['loss_completeness'] = torch.tensor(
+            losses['loss_completeness'], device=completeness_score.device)
```
Comment (dreamerlin) [inline mmaction/models/losses/ssn_loss.py:172]: Redundant

File: mmaction/models/losses/ssn_loss.py [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+
+        if bbox_pred is not None:
+            regression_indexer = (proposal_type == 0).nonzero().squeeze(1)
+            bbox_targets = bbox_targets.view(-1, 2)
+            losses['loss_reg'] = self.classwise_regression_loss(
+                bbox_pred, labels, bbox_targets, regression_indexer)
+            losses['loss_reg'] *= self.loss_weight.reg_loss_weight
+            losses['loss_reg'] = torch.tensor(
+                losses['loss_reg'], device=bbox_pred.device)
```
Comment (dreamerlin) [inline mmaction/models/losses/ssn_loss.py:181]: Redundant.

File: mmaction/models/losses/ohem_hinge_loss.py [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+        ctx.group_size = group_size
+        ctx.num_groups = losses.size(0)
+        return loss
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        labels = ctx.labels
+        slopes = ctx.slopes
+
+        grad_in = torch.zeros(ctx.shape, device=ctx.losses.device)
```
Comment (dreamerlin) [inline mmaction/models/losses/ohem_hinge_loss.py:59]: It seems there are no "ctx.losses", can this code run successfully when training?

Comment (dreamerlin) [reply mmaction/models/losses/ohem_hinge_loss.py:59]: you can write unittest for backward to try it.
